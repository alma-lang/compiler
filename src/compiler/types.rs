use crate::ast::expression::IdentifierName;
use crate::ast::ModuleFullName;
use crate::strings::Strings;
use crate::typ::Types;
use crate::type_env::{PolyTypeEnv, TypeEnv};
use fnv::FnvHashMap;
use std::fmt::Write;

pub type HashMap<K, V> = FnvHashMap<K, V>;

#[derive(Debug)]
pub struct ModuleInterface {
    pub types: TypeEnv,
    pub type_constructors: HashMap<IdentifierName, PolyTypeEnv>,
    pub definitions: PolyTypeEnv,
}

impl ModuleInterface {
    pub fn write_to_string(&self, out: &mut String, strings: &Strings, types: &Types) {
        for (_typ, constructors) in self.type_constructors.iter() {
            constructors.write_to_string(out, strings, types);
            out.push_str("\n\n");
        }

        self.definitions.write_to_string(out, strings, types);
    }
}

#[derive(Debug)]
pub struct ModuleInterfaces(HashMap<ModuleFullName, ModuleInterface>);

impl ModuleInterfaces {
    pub fn new() -> Self {
        let env = HashMap::default();

        Self(env)
    }

    pub fn get(&self, key: &ModuleFullName) -> Option<&ModuleInterface> {
        self.0.get(key)
    }

    pub fn insert(&mut self, key: ModuleFullName, value: ModuleInterface) {
        self.0.insert(key, value);
    }

    pub fn map(&self) -> &HashMap<ModuleFullName, ModuleInterface> {
        &self.0
    }

    pub fn write_to_string(&self, out: &mut String, strings: &Strings, types: &Types) {
        let mut entries: Vec<_> = self.0.iter().collect();
        // we need to sort the entries because they come out with different order and they mess up
        // tests
        entries.sort_by_key(|(k, _)| strings.resolve(**k));

        for (i, (name, interface)) in entries.iter().enumerate() {
            if i > 0 {
                out.push_str("\n\n\n");
            }
            let name = strings.resolve(**name);
            write!(out, "module {name}\n\n").unwrap();
            interface.write_to_string(out, strings, types);
        }
    }
}
