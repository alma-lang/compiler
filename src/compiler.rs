use crate::ast::{self, Module, ReplEntry};
use crate::infer;
use crate::javascript;
use crate::parser;
use crate::source::Source;
use crate::tokenizer;
use crate::type_env::TypeEnv;
use std::collections::HashMap;
use std::rc::Rc;

pub fn compile(source: &Source) -> Result<String, String> {
    let tokens = tokenizer::parse(&source).map_err(|errors| {
        errors
            .iter()
            .map(|e| e.to_string(&source))
            .collect::<Vec<String>>()
            .join("\n\n")
    })?;

    let modules = parser::parse(source, &tokens).map_err(|error| error.to_string(&source))?;

    let mut module_interfaces: HashMap<String, Rc<TypeEnv>> = HashMap::new();
    let mut errors = vec![];

    for module in &modules {
        match infer::infer(&module_interfaces, &module) {
            Ok(typ) => {
                module_interfaces.insert(module.name.to_string(), typ);
            }
            Err((typ, mut module_errors)) => {
                module_interfaces.insert(module.name.to_string(), typ);
                errors.append(&mut module_errors);
            }
        }
    }

    if !errors.is_empty() {
        return Err(errors
            .iter()
            .map(|e| e.to_string(&source))
            .collect::<Vec<String>>()
            .join("\n\n"));
    }

    let javascript_files = javascript::generate(&modules, &module_interfaces);

    let mut out = String::new();

    // Print types
    out.push_str(
        &modules
            .iter()
            .map(|m| {
                format!(
                    "{}\n\n{}\n",
                    m.name.to_string(),
                    module_interfaces
                        .get(&m.name.to_string())
                        .unwrap_or(&Rc::new(TypeEnv::new()))
                )
            })
            .collect::<Vec<String>>()
            .join("\n\n"),
    );

    // Print code
    out.push_str(&javascript::files_to_bundle(&javascript_files));

    Ok(out)
}

pub fn compile_repl_entry(
    module: &mut Module,
    module_interfaces: &mut HashMap<String, Rc<TypeEnv>>,
    source: &Source,
) -> Result<String, String> {
    let tokens = tokenizer::parse(&source).map_err(|errors| {
        errors
            .iter()
            .map(|e| e.to_string(&source))
            .collect::<Vec<String>>()
            .join("\n\n")
    })?;

    let entry = parser::parse_repl(source, &tokens).map_err(|error| error.to_string(&source))?;

    let mut errors = vec![];

    // TODO: Annotate the AST with types. After inference, fetch what we stored in the module and
    // print the type of it
    match entry {
        ReplEntry::Import(import) => {
            module.imports.push(import);
            let result =
                compile_repl_entry_helper(&module, module_interfaces, &source, &mut errors);
            if result.is_err() {
                module.imports.pop();
            }
            result
        }
        ReplEntry::Definition(definition) => {
            module.definitions.push(definition);
            let result =
                compile_repl_entry_helper(&module, module_interfaces, &source, &mut errors);
            if result.is_err() {
                module.definitions.pop();
            }
            result
        }
        ReplEntry::Expression(expression) => {
            module.definitions.push(ast::Definition::Pattern(
                ast::Node {
                    value: ast::Pattern_::Hole,
                    start: 0,
                    end: 0,
                    line: 1,
                    column: 0,
                },
                expression,
            ));
            let result =
                compile_repl_entry_helper(&module, module_interfaces, &source, &mut errors);
            module.definitions.pop();
            result
        }
    }
    .map(|()| "ok.".to_string())
}

// Encapsulate the compiler logic regardless of repl entry type
fn compile_repl_entry_helper<'ast>(
    module: &'ast Module,
    module_interfaces: &mut HashMap<String, Rc<TypeEnv>>,
    source: &Source,
    errors: &mut Vec<infer::Error<'ast>>,
) -> Result<(), String> {
    let result = infer::infer(&module_interfaces, &module);
    match result {
        Ok(typ) => {
            module_interfaces.insert(module.name.to_string(), typ);
        }
        Err((_, mut module_errors)) => {
            errors.append(&mut module_errors);
        }
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors
            .iter()
            .map(|e| e.to_string(&source))
            .collect::<Vec<String>>()
            .join("\n\n"))
    }
}
