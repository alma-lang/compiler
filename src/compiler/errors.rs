pub fn to_string(errors: Vec<String>) -> String {
    errors.join("\n\n----------------------------------------\n\n")
}
