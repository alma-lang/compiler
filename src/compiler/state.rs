use crate::ast::{self, span::Spans, ModuleFullName};
use crate::module_interface::ModuleInterface;
use crate::source::{self, Source, Sources, SourcesData};
use crate::strings::Strings;
use crate::token::Tokens;
use crate::typ::Types;
use derive_more::{From, Into};
use fnv::FnvHashMap;
use typed_index_collections::TiVec;

type HashMap<K, V> = FnvHashMap<K, V>;

#[derive(Debug, From, Into, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
pub struct ModuleIndex(usize);
pub type Modules<T> = TiVec<ModuleIndex, T>;

#[derive(Debug)]
pub struct State {
    pub strings: Strings,
    pub module_to_source_idx: Modules<source::Index>,
    pub module_name_to_module_idx: HashMap<ModuleFullName, ModuleIndex>,
    pub sources: Sources,
    pub tokens: SourcesData<Tokens>,
    pub spans: SourcesData<Spans>,
    pub modules: Modules<ast::Module>,
    pub module_interfaces: Modules<Option<ModuleInterface>>,
    pub types: Types,
}

impl State {
    pub fn new() -> Self {
        Self {
            sources: Sources::new(),
            tokens: SourcesData::new(),
            spans: SourcesData::new(),
            modules: Modules::new(),
            module_interfaces: Modules::new(),
            strings: Strings::new(),
            module_to_source_idx: Modules::new(),
            module_name_to_module_idx: HashMap::default(),
            types: Types::new(),
        }
    }

    pub fn add_source(&mut self, source: Source) -> source::Index {
        let k = self.sources.push_and_get_key(source);
        self.tokens.push(Tokens::new());
        self.spans.push(Spans::new());

        // Return the newly created file index, sources and tokens and spans should always be in
        // sync with file indexes
        debug_assert_eq!(self.sources.len(), self.tokens.len());
        debug_assert_eq!(self.sources.len(), self.spans.len());
        k
    }

    pub fn add_module_ast(
        &mut self,
        source_idx: source::Index,
        module_ast: ast::Module,
    ) -> ModuleIndex {
        let module_name = module_ast.name.full_name;
        let module_idx = self.modules.push_and_get_key(module_ast);
        self.module_interfaces.push(None);
        self.module_to_source_idx.push(source_idx);
        self.module_name_to_module_idx
            .insert(module_name, module_idx);

        // Return the newly created module index, asts and types should always be in sync with
        // module indexes
        debug_assert_eq!(self.modules.len(), self.module_interfaces.len());
        debug_assert_eq!(self.modules.len(), self.module_to_source_idx.len());
        module_idx
    }

    pub fn has_ast(&self, name: ModuleFullName) -> bool {
        self.get_module_idx_with_name(name)
            .map(|_| true)
            .unwrap_or(false)
    }

    pub fn get_module_idx_with_name(&self, name: ModuleFullName) -> Option<ModuleIndex> {
        self.module_name_to_module_idx.get(&name).copied()
    }

    pub fn has_types(&self, name: ModuleFullName) -> bool {
        self.get_types_for_name(name).is_some()
    }

    pub fn get_types_for_name(&self, name: ModuleFullName) -> Option<&ModuleInterface> {
        if let Some(idx) = self.get_module_idx_with_name(name) {
            self.module_interfaces[idx].as_ref()
        } else {
            None
        }
    }

    #[cfg(test)]
    pub fn types_to_string(&self, out: &mut String) {
        use std::fmt::Write;

        for (i, (module_idx, interface)) in self.module_interfaces.iter_enumerated().enumerate() {
            if i > 0 {
                out.push_str("\n\n\n");
            }
            let name_sym = self.modules[module_idx].name.full_name;
            let name = self.strings.resolve(name_sym);
            write!(out, "module {name}\n\n").unwrap();
            interface
                .as_ref()
                .unwrap()
                .write_to_string(out, &self.strings, &self.types);
        }
    }
}
