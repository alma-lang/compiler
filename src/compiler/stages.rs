use crate::ast::ModuleFullName;
use crate::compiler::errors;
use crate::compiler::types::{ModuleAsts, ModuleInterfaces, ModuleSources, Sources};
use crate::infer;
use crate::parser;
use crate::source::Source;
use crate::strings::Strings;
use crate::tokenizer;
use crate::typ::PrimitiveTypes;
use fnv::FnvHashSet as HashSet;

pub fn process_sources(entry_sources: Vec<Source>) -> (Vec<String>, Sources) {
    let entry_files: Vec<String> = entry_sources.iter().map(|s| s.name().to_string()).collect();

    let mut sources = Sources::default();
    for source in entry_sources {
        sources.insert(source.name().to_string(), source);
    }

    (entry_files, sources)
}

pub fn parse_files<'sources>(
    entry_files: &[String],
    sources: &'sources Sources,
    strings: &mut Strings,
) -> Result<(Vec<ModuleFullName>, ModuleSources<'sources>, ModuleAsts), String> {
    let mut errors: Vec<String> = vec![];
    let mut module_sources = ModuleSources::default();
    let mut module_asts = ModuleAsts::default();
    let mut entry_modules = vec![];

    for file in entry_files {
        let source = sources.get(file).unwrap();

        let mut parse_queue: Vec<ModuleFullName> = vec![];

        let parse_result = parse_file(
            source,
            &mut parse_queue,
            &mut module_asts,
            &mut module_sources,
            strings,
        );

        match parse_result {
            Ok(modules) => {
                entry_modules.extend(modules);

                loop {
                    match parse_queue.pop() {
                        None => break,
                        Some(module_name) => {
                            if !module_asts.contains_key(&module_name) {
                                // Unparsed module.

                                // Resolve module to file path and parse it or error
                                // Maybe not error since infer will error with the import not
                                // found?
                                // TODO
                                // parse_file(&resolved_source, &mut parse_queue, &mut module_asts)?;
                            }
                        }
                    }
                }
            }
            Err(module_errors) => errors.extend(module_errors),
        }
    }

    if errors.is_empty() {
        Ok((entry_modules, module_sources, module_asts))
    } else {
        Err(errors::to_string(errors))
    }
}

fn parse_file<'source>(
    source: &'source Source,
    parse_queue: &mut Vec<ModuleFullName>,
    module_asts: &mut ModuleAsts,
    module_sources: &mut ModuleSources<'source>,
    strings: &mut Strings,
) -> Result<Vec<ModuleFullName>, Vec<String>> {
    let tokens = tokenizer::parse(source).map_err(|errors| {
        errors
            .iter()
            .map(|e| e.to_string(source))
            .collect::<Vec<String>>()
    })?;

    let mut module_names = vec![];

    let (module, submodules) =
        parser::parse(source, &tokens, strings).map_err(|error| vec![error.to_string(source)])?;

    for module in submodules {
        // Add all submodules to queue
        parse_queue.push(module.name.full_name);
        module_names.push(module.name.full_name);
        module_sources.insert(module.name.full_name, source);

        // Queue dependencies
        let dependencies = module.dependencies();
        for dependency in dependencies {
            let name = &dependency.full_name;
            if !module_asts.contains_key(name) {
                parse_queue.push(*name);
            }
        }

        module_asts.insert(module.name.full_name, module);
    }
    // And put the top level one last to start with it
    parse_queue.push(module.name.full_name);
    module_names.push(module.name.full_name);
    module_sources.insert(module.name.full_name, source);
    // Queue dependencies
    let dependencies = module.dependencies();
    for dependency in dependencies {
        let name = &dependency.full_name;
        if !module_asts.contains_key(name) {
            parse_queue.push(*name);
        }
    }
    module_asts.insert(module.name.full_name, module);

    Ok(module_names)
}

pub fn check_cycles(
    entry_modules: &[ModuleFullName],
    module_asts: &ModuleAsts,
    strings: &Strings,
) -> Result<Vec<ModuleFullName>, String> {
    let mut sorted_modules = vec![];
    let mut errors: Vec<String> = vec![];
    let mut visited_modules: HashSet<ModuleFullName> = HashSet::default();
    let mut entry_points_queue: Vec<Vec<ModuleFullName>> = vec![];
    entry_points_queue.extend(entry_modules.iter().map(|m| vec![*m]));

    fn last_visited_is(name: &ModuleFullName, visited_path: &[ModuleFullName]) -> bool {
        visited_path.last().map(|m| m == name).unwrap_or(false)
    }

    // For each entry point module, check for cycles separately
    'check_from_module: loop {
        let mut visited_path: Vec<ModuleFullName> = Vec::new();

        match entry_points_queue.pop() {
            None => break,
            Some(mut modules_queue) => loop {
                // Check the modules to visit queue
                match modules_queue.pop() {
                    None => break,
                    Some(module_name) => {
                        let already_visited = visited_modules.contains(&module_name);

                        // If I haven't checked this module before at all in any entrypoint...
                        if !already_visited {
                            // Place it in the queue of modules to check it, in the visited modules
                            // path, and add its dependencies to check after it to the modules
                            // queue
                            let module = module_asts.get(&module_name).unwrap();
                            modules_queue.push(module_name);
                            visited_modules.insert(module_name);
                            visited_path.push(module_name);
                            sorted_modules.push(module_name);

                            if !module.imports.is_empty() {
                                for module in module.dependencies() {
                                    // Check for direct cycles before adding it to the queue since
                                    // this disrupts the popping of elements from visited_path when
                                    // the module directly references itself, because we check by
                                    // name. Instead report the error here if it exists and bail
                                    // out.
                                    if module.full_name != module_name {
                                        modules_queue.push(module.full_name);
                                    } else {
                                        let name = strings.resolve(module_name);
                                        errors.push(cycle_error(&[name, name]));
                                        break 'check_from_module;
                                    }
                                }
                            }
                        } else {
                            // If the module has already been visited then we check it for cycles
                            // in the current visited_path

                            // If the module was visited during this modules queue, it should be in
                            // the visited_path last, so we remove it for check_cycles_in_path.
                            // This one check was problematic with direct cycles like A imports A,
                            // but this is checked separately when adding dependencies above.
                            if last_visited_is(&module_name, &visited_path) {
                                visited_path.pop();
                            }

                            let found_cycle = check_cycles_in_path(
                                &module_name,
                                &visited_path,
                                &mut errors,
                                strings,
                            );

                            // If there is a cycle, don't bother checking anything else for this
                            // entry point module, try reporting cycles in other entry points.
                            if found_cycle {
                                break 'check_from_module;
                            }
                        }
                    }
                }
            },
        }
    }

    if errors.is_empty() {
        Ok(sorted_modules)
    } else {
        Err(errors::to_string(errors))
    }
}

fn check_cycles_in_path(
    module_name: &ModuleFullName,
    visited_path: &[ModuleFullName],
    errors: &mut Vec<String>,
    strings: &Strings,
) -> bool {
    let mut cycle = false;
    let mut cycle_names = visited_path;

    for (i, visited) in visited_path.iter().rev().enumerate() {
        if visited == module_name {
            cycle = true;
            cycle_names = &visited_path[visited_path.len() - (i + 1)..];
            break;
        }
    }

    if cycle {
        let mut cycle_names: Vec<&str> = cycle_names.iter().map(|m| strings.resolve(*m)).collect();
        cycle_names.push(strings.resolve(*module_name));

        errors.push(cycle_error(&cycle_names));
    }

    cycle
}

fn cycle_error(cycle_names: &[&str]) -> String {
    format!(
        "Cycle detected between modules:\n\n    {}\n\n\
                            Alma's module system does not support \
                            cyclic dependencies.",
        cycle_names.join("\n        â†“\n    ")
    )
}

pub fn infer(
    entry_modules: Vec<ModuleFullName>,
    module_sources: &ModuleSources,
    module_asts: &ModuleAsts,
    primitive_types: &PrimitiveTypes,
    strings: &mut Strings,
) -> Result<ModuleInterfaces, String> {
    let mut errors = vec![];
    let mut module_interfaces = ModuleInterfaces::new();
    let mut infer_queue: Vec<ModuleFullName> = vec![];
    infer_queue.extend(entry_modules);

    loop {
        match infer_queue.pop() {
            None => break,
            Some(module_name) => {
                // If not already inferred
                if module_interfaces.map().contains_key(&module_name) {
                    continue;
                }
                let module = module_asts.get(&module_name).unwrap_or_else(|| {
                    panic!(
                        "Couldn't find module {} in the ASTs",
                        strings.resolve(module_name)
                    )
                });

                // Check dependencies and skip back to queue if there are any that need processing
                // before this module.
                {
                    let dependencies = module.dependencies();
                    let mut skip = false;
                    for dependency in dependencies {
                        let name = &dependency.full_name;
                        // If not inferred but parsed, ready for inference.
                        // If not parsed, it means it couldn't be found, and the module who
                        // imported it will report the error. Don't add it to the infer queue.
                        if !module_interfaces.map().contains_key(name)
                            && module_asts.contains_key(name)
                        {
                            if !skip {
                                // At least one dependency to be proccessed first,
                                // put this module first in the queue and then the deps
                                skip = true;
                                infer_queue.push(module_name);
                            }
                            infer_queue.push(*name);
                        }
                    }
                    if skip {
                        continue;
                    }
                }

                // Process this module
                let typ = match infer::infer(&module_interfaces, module, primitive_types, strings) {
                    Ok(typ) => typ,
                    Err((typ, module_errors)) => {
                        errors.append(
                            &mut module_errors
                                .iter()
                                .map(|e| {
                                    e.to_string(module_sources.get(&module_name).unwrap(), strings)
                                })
                                .collect::<Vec<String>>(),
                        );
                        typ
                    }
                };

                module_interfaces.insert(module_name, typ);
            }
        };
    }

    if errors.is_empty() {
        Ok(module_interfaces)
    } else {
        Err(errors::to_string(errors))
    }
}
