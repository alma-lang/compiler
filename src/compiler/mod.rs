pub mod errors;
pub mod stages;
pub mod types;

use crate::ast::{self, Module, ReplEntry};
use crate::compiler::stages::{check_cycles, infer, parse_files, process_sources};
use crate::compiler::types::ModuleInterfaces;
use crate::infer;
use crate::javascript;
use crate::parser;
use crate::source::Source;
use crate::tokenizer;
use std::fmt::Write;

pub fn compile(entry_sources: Vec<Source>) -> Result<String, String> {
    let (entry_sources, sources) = process_sources(entry_sources);

    let (entry_modules, module_sources, module_asts) = parse_files(&entry_sources, &sources)?;

    check_cycles(&entry_modules, &module_asts)?;

    let module_interfaces = infer(entry_modules, &module_sources, &module_asts)?;

    let javascript_files = javascript::generate(&module_asts, &module_interfaces);

    let mut out = String::new();

    // Print types
    write!(&mut out, "{}", &module_interfaces).unwrap();

    // Print code
    out.push_str(&javascript::files_to_bundle(&javascript_files));

    Ok(out)
}

pub fn compile_repl_entry(
    module: &mut Module,
    module_interfaces: &mut ModuleInterfaces,
    source: &Source,
) -> Result<String, String> {
    let tokens = tokenizer::parse(source).map_err(|errors| {
        errors
            .iter()
            .map(|e| e.to_string(source))
            .collect::<Vec<String>>()
            .join("\n\n")
    })?;

    let entry = parser::parse_repl(source, &tokens).map_err(|error| error.to_string(source))?;

    let mut errors = vec![];

    // TODO: Annotate the AST with types. After inference, fetch what we stored in the module and
    // print the type of it
    match entry {
        ReplEntry::Import(import) => {
            module.imports.push(import);
            let result = compile_repl_entry_helper(module, module_interfaces, source, &mut errors);
            if result.is_err() {
                module.imports.pop();
            }
            result
        }
        ReplEntry::Definition(definition) => {
            module.definitions.push(definition);
            let result = compile_repl_entry_helper(module, module_interfaces, source, &mut errors);
            if result.is_err() {
                module.definitions.pop();
            }
            result
        }
        ReplEntry::Expression(expression) => {
            module.definitions.push(ast::Definition::Pattern(
                ast::Node {
                    value: ast::Pattern_::Hole,
                    start: 0,
                    end: 0,
                    line: 1,
                    column: 0,
                },
                expression,
            ));
            let result = compile_repl_entry_helper(module, module_interfaces, source, &mut errors);
            module.definitions.pop();
            result
        }
    }
    .map(|()| "ok.".to_string())
}

// Encapsulate the compiler logic regardless of repl entry type
fn compile_repl_entry_helper<'ast>(
    module: &'ast Module,
    module_interfaces: &mut ModuleInterfaces,
    source: &Source,
    errors: &mut Vec<infer::Error<'ast>>,
) -> Result<(), String> {
    let result = infer::infer(module_interfaces, module);
    match result {
        Ok(typ) => {
            module_interfaces.insert(module.name.full_name.clone(), typ);
        }
        Err((_, mut module_errors)) => {
            errors.append(&mut module_errors);
        }
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors
            .iter()
            .map(|e| e.to_string(source))
            .collect::<Vec<String>>()
            .join("\n\n"))
    }
}

#[cfg(test)]
pub mod tests {
    use crate::source::Source;
    use insta::assert_snapshot;

    #[test]
    pub fn test_compile() {
        fn compile(source_codes: &[&'static str]) -> String {
            match super::compile(
                source_codes
                    .iter()
                    .map(|s| Source::new_orphan(s.to_string()))
                    .collect(),
            ) {
                Ok(s) => s,
                Err(e) => e,
            }
        }

        // Normal modules

        assert_snapshot!(compile(&[r"
module Modules exposing (main)

import Modules.WeirdMath exposing (weirdAdd)

main = weirdAdd



module Modules.Constants exposing (five)

    five = 5



module Modules.WeirdMath exposing (weirdAdd)

    import Modules.Constants exposing (five)

    weirdAdd = \x y -> x * y + five
        "]));

        // Cycle modules

        assert_snapshot!(compile(&[r"
module CycleModule

import CycleModule.Test


module CycleModule.Test
    import CycleModule
        "]));

        assert_snapshot!(compile(&[r"
module CycleModule

import CycleModule
        "]));
    }
}
