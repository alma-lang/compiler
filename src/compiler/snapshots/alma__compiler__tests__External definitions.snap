---
source: src/compiler/mod.rs
assertion_line: 253
expression: "compile(&[r\"\nmodule Test exposing (main)\n\nexternal test : Float -> Float -> Float\n\nmain a = test a 5\n\nmodule Test.Inner exposing (another_test)\n    external another_test : Float -> Float\n        \"])"
---
// alma_out/Test.Inner.js

/*

module Test.Inner

another_test : ∀ ∅ . Float -> Float

*/


import * as Test__Inner__ffi from "alma_out/Test.Inner.ffi.js"


import * as Alma from "alma_out/Alma.js"
import { True, False } from "alma_out/Alma.js"

let another_test = Test__Inner__ffi.another_test

export {
    another_test,
}



// alma_out/Alma.js

/*

module Alma



False : ∀ ∅ . Bool

True : ∀ ∅ . Bool



add : ∀ ∅ . Float -> Float -> Float

and_ : ∀ ∅ . Bool -> Bool -> Bool

div : ∀ ∅ . Float -> Float -> Float

eq : ∀ a . a -> a -> Bool

ge : ∀ ∅ . Float -> Float -> Bool

gt : ∀ ∅ . Float -> Float -> Bool

le : ∀ ∅ . Float -> Float -> Bool

lt : ∀ ∅ . Float -> Float -> Bool

mult : ∀ ∅ . Float -> Float -> Float

ne : ∀ a . a -> a -> Bool

or_ : ∀ ∅ . Bool -> Bool -> Bool

sub : ∀ ∅ . Float -> Float -> Float

*/





// type Bool

let True = {
    tag: "True",
}

let False = {
    tag: "False",
}


let or_ = new Error("Unimplemented")

let and_ = new Error("Unimplemented")

let eq = new Error("Unimplemented")

let ne = new Error("Unimplemented")

let gt = new Error("Unimplemented")

let ge = new Error("Unimplemented")

let lt = new Error("Unimplemented")

let le = new Error("Unimplemented")

let add = new Error("Unimplemented")

let sub = new Error("Unimplemented")

let mult = new Error("Unimplemented")

let div = new Error("Unimplemented")

export {
    // type Bool
    True,
    False,
    or_,
    and_,
    eq,
    ne,
    gt,
    ge,
    lt,
    le,
    add,
    sub,
    mult,
    div,
}



// alma_out/Test.js

/*

module Test

main : ∀ ∅ . Float -> Float

*/


import * as Test__ffi from "alma_out/Test.ffi.js"


import * as Alma from "alma_out/Alma.js"
import { True, False } from "alma_out/Alma.js"

let test = Test__ffi.test

function main(a) {
    return test(a, 5)
}

export {
    main,
}




